grinpy/invariants/zero_forcing.py

```diff
@@ -93,6 +93,16 @@ def is_zero_forcing_set(G, S):
     # TODO: Add documentation
     return is_k_forcing_set(G, S, 1)
 
+def is_total_forcing_set(G, S):
+    # TODO: Add documentation
+    no_isolates = True
+    for v in S:
+        if neighborhood(G, v).intersection(S) == {}:
+           no_isolates = False
+           return False
+    return is_zero_forcing_set(G, S)
```

Comment:
This will set off an infinite loop. I think you want to return `no_isolates`. You can clean this up, though:

```[python]
def is_total_forcing_set(G, S)
  for v in S:
    if set(neighborhood(G, v)).intersection(S) == {}:
      return False
  return True
```
Note the `set()` wrapper around `neighborhood` because `neighborhood` returns a list, not a set.

Also, any time you expect an iterable as an argument you should check that it is actually iterable and either force it into one or throw an exception if it isn't. In this case, we can just force it into a set if it isn't already, so the full function would look something like this:

```[python]
def is_total_forcing_set(G, S)
  try:
    _ = (v for v in S)
  except:
    S = [S] # forces S into a list
  # now that we know S is iterable, we can force it into a set
  S = set(S)
  for v in S:
    if set(neighborhood(G, v)).intersection(S) == {}:
      return False
  return True
```
It may seem long winded, but it makes sure the function behaves the way we expect it to.

Reply:
@RandyRDavila , I just realized that you were returning `is_zero_forcing_set(G, S)` and not `is_total_forcing_set(G, S)`, so this will _not_ enter into an infinite loop. Your code is correct, but the `no_isolates` boolean is not needed. I've made the changes in my latest commit.